src/heap_utils.cc

```diff
@@ -448,30 +448,60 @@ void CreateHeapSnapshotStream(const FunctionCallbackInfo<Value>& args) {
 void TriggerHeapSnapshot(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
   Isolate* isolate = args.GetIsolate();
-  CHECK_EQ(args.Length(), 2);
+  CHECK_EQ(args.Length(), 3);
   Local<Value> filename_v = args[0];
   auto options = GetHeapSnapshotOptions(args[1]);
+  Local<String> path_v = args[2].As<String>();
+  bool has_path = !path_v->IsNullOrUndefined() && path_v->IsString();
+  std::string final_filename;
 
   if (filename_v->IsUndefined()) {
     DiagnosticFilename name(env, "Heap", "heapsnapshot");
-    THROW_IF_INSUFFICIENT_PERMISSIONS(
-        env,
-        permission::PermissionScope::kFileSystemWrite,
-        Environment::GetCwd(env->exec_path()));
-    if (WriteSnapshot(env, *name, options).IsNothing()) return;
-    if (String::NewFromUtf8(isolate, *name).ToLocal(&filename_v)) {
-      args.GetReturnValue().Set(filename_v);
+    if (!has_path) {
+      THROW_IF_INSUFFICIENT_PERMISSIONS(
+          env,
+          permission::PermissionScope::kFileSystemWrite,
+          Environment::GetCwd(env->exec_path()));
+      final_filename = *name;
+    } else {
+      BufferValue path(isolate, path_v);
+      CHECK_NOT_NULL(*path);
+      ToNamespacedPath(env, &path);
+      THROW_IF_INSUFFICIENT_PERMISSIONS(
+          env,
+          permission::PermissionScope::kFileSystemWrite,
+          path.ToStringView());
+
+      final_filename = std::string(*path) + "/" +
+                       *name;  // NOLINT(readability/pointer_notation)
+    }
+  } else {
+    BufferValue filename(isolate, filename_v);
+    CHECK_NOT_NULL(*filename);
+
+    if (has_path) {
+      BufferValue path(isolate, path_v);
+      CHECK_NOT_NULL(*path);
+      ToNamespacedPath(env, &path);
+      THROW_IF_INSUFFICIENT_PERMISSIONS(
+          env,
+          permission::PermissionScope::kFileSystemWrite,
+          std::string(*path) + "/" + std::string(*filename));
+      final_filename = std::string(*path) + "/" + std::string(*filename);
```

Comment:
```suggestion
      final_filename = path.ToString() + "/" + filename.ToString();
```

But I'd encourage you to move this line up before the `THROW_IF_INSUFFICIENT_PERMISSIONS` one and then just use the same variable in that

Reply:
So true! Thanks!